docs/dev/topology-over-raft.md

```diff
@@ -224,3 +224,64 @@ There are also a few static columns for cluster-global properties:
 - `unpublished_cdc_generations` - the IDs of the committed yet unpublished CDC generations
 - `global_topology_request` - if set, contains one of the supported global topology requests
 - `new_cdc_generation_data_uuid` - used in `commit_cdc_generation` state, the time UUID of the generation to be committed
+
+# Join procedure
+
+In topology on raft mode, new nodes need to go through a new handshake procedure
+before they can join the cluster, including joining group 0. The handshake
+happens during raft discovery and replaced the old `GROUP0_MODIFY_CONFIG`.
+
+Two RPC verbs are introduced:
+
+- `JOIN_NODE_REQUEST`
+  Sent by the node that wishes to join the cluster. It contains some basic
+  information about the node that will have to be verified. Some of them
+  can be verified by the node that receives the RPC - which can result in
+  an immediate failure of the RPC -  but others need to be verified
+  by the topology coordinator.
+
+- `JOIN_NODE_REPLY`
+  Sent by the topology coordinator back to the joining node. It contains
+  coordinator's decision about the node - whether the request was accepted
+  or rejected.
+
+The procedure is not retryable. If the joining node crashes before finishing it,
+it might get rejected after restart. In order to retry adding the node, the data
+directory must be deleted first.
+
+*The procedure*
+
+If the node didn't join group 0, it sends `JOIN_NODE_REQUEST` to any existing
+node in the cluster. The receiving node can either:
+- Accept the request and tell the new node to wait for `JOIN_NODE_RESPONSE`.
+- Reject the request. This can happen if:
+  - The request does not satisfy some validity checks done by the receiving node
+    (e.g. cluster name does not match),
+  - There already an existing request to join this node (which can happen if
+    the node crashed during the previous attempt to join),
+  - The node is already a part of the cluster,
+  - The node was removed from the cluster.
+
+The new node should not process `JOIN_NODE_RESPONSE` RPC until it is explicitly
+ordered by node that handles `JOIN_NODE_REQUEST`. This is necessary so that
+it doesn't accidentally process a join request from a previous attempt, if there
```

Comment:
But can this happen after we made the procedure not restartable?

Reply:
Yes, the new node uses `JOIN_NODE_REQUEST` to check whether it previously tried to join - the node that handles the request will reject it in such case.

I suppose the joining node could store a persistent flag which tells whether it successfully finished joining or not. I don't think it would simplify the implementation much (only that group0 handshaker would have one method instead of two) but perhaps would simplify reasoning about the procedure and this paragraph would be unnecessary.

Reply:
I try to remove flags, not add them If the situation may indeed happen I am fine with the way it is handled.